src/util-inl.h

```diff
@@ -401,6 +401,29 @@ inline char* UncheckedCalloc(size_t n) { return UncheckedCalloc<char>(n); }
 // headers than we really need to.
 void ThrowErrStringTooLong(v8::Isolate* isolate);
 
+struct ArrayIterationData {
+  std::vector<v8::Global<v8::Value>>* out;
```

Comment:
Can we use `std::vector<v8::Local<v8::Value>>` for the out param? As far as I can tell, the usage sites above only use the handles in their local scope.

Reply:
I've [tried that before](https://github.com/joyeecheung/node/tree/array-iteration-bench) but saw that it would be always be slower than just returning global handles (the API only allows creating global handles during the iteration).

Reply:
Numbers I got from the benchmarks, where `SumWithTemporaryGlobals` (summing the elements in an array with a temporary `std::vector<v8::Local<v8::Value>>` vector created from a `std::vector<v8::Global<v8::Value>>` vector which is necessary during API restriction) is always slower than `SumWithGlobals` (where the locals are only created during the actual iteration where operations are performed on them).

```
misc/sum.js kind="SumWithGlobals" length=1 n=10000: 9,837,523.462493457
misc/sum.js kind="SumTraditional" length=1 n=10000: 9,120,750.528319474
misc/sum.js kind="SumWithTemporaryGlobals" length=1 n=10000: 7,025,032.2975859875
misc/sum.js kind="SumByNative" length=1 n=10000: 11,288,784.028628357
misc/sum.js kind="SumWithGlobals" length=2 n=10000: 7,102,237.417853747
misc/sum.js kind="SumTraditional" length=2 n=10000: 5,881,827.03664143
misc/sum.js kind="SumWithTemporaryGlobals" length=2 n=10000: 6,772,259.740541185
misc/sum.js kind="SumByNative" length=2 n=10000: 9,904,059.376816776
misc/sum.js kind="SumWithGlobals" length=3 n=10000: 5,711,335.40159826
misc/sum.js kind="SumTraditional" length=3 n=10000: 4,950,901.90580018
misc/sum.js kind="SumWithTemporaryGlobals" length=3 n=10000: 5,126,107.367344031
misc/sum.js kind="SumByNative" length=3 n=10000: 8,694,729.776493277
misc/sum.js kind="SumWithGlobals" length=10 n=10000: 2,750,912.0648951163
misc/sum.js kind="SumTraditional" length=10 n=10000: 2,131,967.5096679395
misc/sum.js kind="SumWithTemporaryGlobals" length=10 n=10000: 2,506,418.9389025294
misc/sum.js kind="SumByNative" length=10 n=10000: 3,335,821.856438236
misc/sum.js kind="SumWithGlobals" length=100 n=10000: 365,865.06728020776
misc/sum.js kind="SumTraditional" length=100 n=10000: 248,260.23571365993
misc/sum.js kind="SumWithTemporaryGlobals" length=100 n=10000: 356,812.29603717127
misc/sum.js kind="SumByNative" length=100 n=10000: 821,605.058063651
```

Reply:
Hmm, I see the differences here. In this sense, maybe it can be more straightforward to iterate the array in place instead of creating a new vector for the elements: https://github.com/legendecas/node/commit/6becdac35e5682627359892ded5440078b1cd466

The improvements can be significant with in-place iteration when the length grows:

```
misc/sum.js kind="SumWithIteration" length=1 n=10000: 14,854,250.098038051
misc/sum.js kind="SumWithGlobals" length=1 n=10000: 15,177,385.695313983
misc/sum.js kind="SumTraditional" length=1 n=10000: 17,488,907.660316445
misc/sum.js kind="SumWithTemporaryGlobals" length=1 n=10000: 13,108,307.389808292
misc/sum.js kind="SumByNative" length=1 n=10000: 24,188,652.13573704
misc/sum.js kind="SumWithIteration" length=2 n=10000: 24,330,900.243309002
misc/sum.js kind="SumWithGlobals" length=2 n=10000: 13,645,655.08696376
misc/sum.js kind="SumTraditional" length=2 n=10000: 12,274,954.060984427
misc/sum.js kind="SumWithTemporaryGlobals" length=2 n=10000: 11,527,377.521613833
misc/sum.js kind="SumByNative" length=2 n=10000: 20,061,830.56179144
misc/sum.js kind="SumWithIteration" length=3 n=10000: 22,257,240.28026317
misc/sum.js kind="SumWithGlobals" length=3 n=10000: 11,762,976.032936333
misc/sum.js kind="SumTraditional" length=3 n=10000: 9,929,254.064788383
misc/sum.js kind="SumWithTemporaryGlobals" length=3 n=10000: 9,752,929.048416466
misc/sum.js kind="SumByNative" length=3 n=10000: 18,069,589.603480924
misc/sum.js kind="SumWithIteration" length=10 n=10000: 16,909,744.24011837
misc/sum.js kind="SumWithGlobals" length=10 n=10000: 6,026,365.348399246
misc/sum.js kind="SumTraditional" length=10 n=10000: 3,828,055.975365694
misc/sum.js kind="SumWithTemporaryGlobals" length=10 n=10000: 5,396,290.050590219
misc/sum.js kind="SumByNative" length=10 n=10000: 8,003,201.280512205
misc/sum.js kind="SumWithIteration" length=100 n=10000: 3,507,439.1029654345
misc/sum.js kind="SumWithGlobals" length=100 n=10000: 819,034.3584913387
misc/sum.js kind="SumTraditional" length=100 n=10000: 453,651.04027854174
misc/sum.js kind="SumWithTemporaryGlobals" length=100 n=10000: 748,302.3078466307
misc/sum.js kind="SumByNative" length=100 n=10000: 1,407,376.935090509
```

Reply:
I think in-place iteration can be very limited in practice, because the callback disallows heap allocation. For example for the Blob here we can't do the operation in-place because that involves hep allocation. The majority of migration targets I can find in the code base can't do that either.

Reply:
Thank you for the clarification!